types/types.cc

```diff
@@ -3679,11 +3679,17 @@ make_user_value(data_type type, user_type_impl::native_type value) {
     return data_value::make_new(std::move(type), std::move(value));
 }
 
-std::ostream& operator<<(std::ostream& out, const data_value& v) {
+ auto fmt::formatter<data_value>::format(const data_value& v,
+                                         fmt::format_context& ctx) const -> decltype(ctx.out()) {
     if (v.is_null()) {
-        return out << "null";
+        return fmt::format_to(ctx.out(), "null");
     }
-    return out << v.type()->to_string_impl(v);
+    return fmt::format_to(ctx.out(), "{}", v.type()->to_string_impl(v));
+}
+
+std::ostream& operator<<(std::ostream& out, const data_value& v) {
+    fmt::print(out, "{}", v);
+    return out;
```

Comment:
I wonder if we can't add a generic (template) operator<< which can output anything using fmt::print(), to the single place you said needed it (the place where we format a vector).

Reply:
in this very case, there are at least two places i ran into. one is in the test, another is in production code. but if we are going to add a generic operator<< for the types which does not have `operator<<`, we would have to put it into the `std` namespace, as the types could belong to different namespaces, and accordingly to how ADT works, we would need to put the operator into one of its argument, and this is not quite right, as we don't own the `std` namespace. and IMHO, we would be in the business of `utils/to_string.hh`, and it's a mistake.

Reply:
It's tempting but it's better to work towards removing operator<< completely.